### PR TITLE
Oauth2 granular scopes

### PR DESCRIPTION
**Pull request by marcellmars**
## Enhancing Gitea OAuth2 Provider with Granular Scopes for Resource Access

This PR was initiated following my personal research to find the lightest possible Single Sign-On solution for self-hosted setups. The existing solutions often seemed too enterprise-oriented, involving many moving parts and services, demanding significant resources while promising planetary-scale capabilities. Others were adequate in supporting basic OAuth2 flows but lacked proper user management features, such as a change password UI.

Gitea hits the sweet spot for me, provided it supports more granular access permissions for resources under users who accept the OAuth2 application.

This PR aims to introduce this capability as nonintrusively and simply as possible. It acknowledges that the default scope is `all` and begins to process only additional scopes typical for OIDC/OAuth2: openid, profile, email, and groups. These then should be the ones already established with personal tokens.

Personal tokens define scopes around specific resources: user info, repositories, issues, packages, organizations, notifications, miscellaneous, admin, and activitypub, with access delineated by read and/or write permissions.

The initial case I wanted to address was to have Gitea act as an OAuth2 Identity Provider. To achieve that, with this PR, I would only add `public-only` and have no access to any API endpoint. The rest of the scopes should be provided as it is now, combining the existing Gitea OAuth2 scopes: openid, profile, email, and groups.

Another example: if a third party wanted to interact solely with Issues, it would need to add  `read:issue`/`write:issue` to manage Issues.

Another scenario could involve adding `read:user` and `public-only`, preventing the third party from accessing anything the user has kept private.

My approach is based on my understanding of how scopes can be utilized, supported by examples like [Sample Use Cases: Scopes and Claims](https://auth0.com/docs/get-started/apis/scopes/sample-use-cases-scopes-and-claims) on auth0.com.

I added an additional check to `CheckOAuthAccessToken` that returns `AccessTokenScopeAll` if no additional scopes or `AccessTokenScope` with whatever requested instead.  That `AccessTokenScope` is then assigned to `store.GetData()["ApiTokenScope"]` in `userIDFromToken` to be checked against the accepted/stored grant scopes. The main difference is the opportunity to reduce the permissions from `all`, as is currently the case, to what is provided by the additional scopes described above.

The intent of this PR is to ensure that if nothing is added, or if scopes are added to `openid, profile, email, groups` but not in `AccessTokenScope`, everything continues to work as it currently does.

### CODE REVIEW

**Comment by marcellmars** [comment]
i can see all the tests failing at the moment in the `test-sqlite` suite are related to asking for more scopes than before. i think we need a policy when requesting per token resources nested inside others. for example, if a token has only `[read:organization]` but `repos` are requested in the route path **after** the organization. this might differ if it's nested under `admin` or under `org`.

i can compile an exhaustive list with questions, but if there's already a policy, it would be great to know it in advance.

my current questions are:
- if we have a path starting with `/org`, do we need to ask for anything else under `/org`? do we only care about the root resources when creating a token?

if we say "yes" to the root resource/path, i can imagine someone might not want to share repositories via organizations if those repositories are already restricted (and no token exists for that).

i imagine we could take a granular approach: if you want repositories under an organization, for example, the token would need both `[read:organization]` and `[read:repository]`. but we could change this for admin so that `[read:admin]` would allow access to everything under admin.

i'm aware that some of these changes would be breaking ones. one way to address this issue is to first ensure all previous tests are passing and then see what new issues arise.

those changes would mostly come from the revisited `tokenRequiresScopes` middleware under all the routes/paths in `routers/api/v1/api.go`

File: modules/setting/oauth2.go [hunk 1]
```diff
@@ -90,23 +90,25 @@ func parseScopes(sec ConfigSection, name string) []string {
...
-	DefaultApplications        []string
+	Enabled                     bool
+	AccessTokenExpirationTime   int64
+	RefreshTokenExpirationTime  int64
+	InvalidateRefreshTokens     bool
+	JWTSigningAlgorithm         string `ini:"JWT_SIGNING_ALGORITHM"`
+	JWTSigningPrivateKeyFile    string `ini:"JWT_SIGNING_PRIVATE_KEY_FILE"`
+	MaxTokenLength              int
+	DefaultApplications         []string
+	EnableAdditionalGrantScopes bool
```
Comment (lunny) [inline modules/setting/oauth2.go:101]: Do we really need this configuration item?

**Comment by techknowlogick** [comment]
Apologies if this is out of scope, but it might be useful to give a description next to the used scopes as a plain language would likely lower any chance at additional scopes are accepted mistakenly (def out of scope for this PR is including a warning if admin scope is requested)

File: modules/setting/oauth2.go (hunk 1 above)
Comment (marcellmars) [reply modules/setting/oauth2.go:101]: we don't need the configuration item.

it was introduced with the intention of giving us some time to test it with opt-in enthusiasts and also for those who wanted to try it with their existing tokens. if it didn’t work for them, they could easily revert to the previous setup.

my attempts to gauge how much personal access tokens and oauth2 are being used (by asking in chat channels and searching issues for both gitea and forgejo) led me to the conclusion of "not that much"... though i realize i might be way off here :)

i'll commit the changes with the configuration item removed. if you decide to revert, it will be easy to do so.

**Comment by marcellmars** [comment]
> Apologies if this is out of scope, but it might be useful to give a description next to the used scopes as a plain language would likely lower any chance at additional scopes are accepted mistakenly (def out of scope for this PR is including a warning if admin scope is requested)

are you thinking of adding plain language descriptions as code comments, or user interface changes that would warn users while they approve the required scopes?

currently, users are warned in the web ui about full access to all resources. are you suggesting we provide a clearer explanation of what happens when someone grants more limited or restrictive access?

File: modules/setting/oauth2.go (hunk 1 above)
Comment (marcellmars) [reply modules/setting/oauth2.go:101]: resolving with f50b027380c21bd343b6fade578ce9b60af548f5

**Comment by marcellmars** [comment]
I added a section at the top of the description. Pasted here too:

## Required scopes per API call

There are two possible approaches for determining the required scopes for any API call:

1. The scope required is determined by the root of `/api/v1/` (e.g., `/api/v1/user/repos` would require only `read:user` to read repositories).
2. The required scopes follow the entire route/path, so accessing `/api/v1/user/repos` would require both `read:user` AND `read:repository`.

Before this PR, neither of these approaches was consistently applied. Instead, there was a mixture of both.

I suggest implementing the second approach, which requires listing the scope for every part of the path. One of the main reasons for this recommendation is to fulfill the promise of `public-only` repositories and organizations introduced with personal access tokens. This is understood to mean that no private repositories should be exposed if a `public-only` scope is added to the token making an API call.

The last commit "make old tests work with granular scopes" changed the scopes for the older tests in order to work with the changes.

File: templates/user/auth/grant.tmpl [hunk 2]
```diff
@@ -8,8 +8,11 @@
 			<div class="ui attached segment">
 				{{template "base/alert" .}}
 				<p>
+					{{if not .AdditionalScopes}}
 					<b>{{ctx.Locale.Tr "auth.authorize_application_description"}}</b><br>
+					{{end}}
 					{{ctx.Locale.Tr "auth.authorize_application_created_by" .ApplicationCreatorLinkHTML}}
+				<p>With scopes: {{.Scope}}.</p>
```
Comment (lunny) [inline templates/user/auth/grant.tmpl]: Translation is necessary.

File: routers/web/auth/oauth2_provider.go [hunk 3]
```diff
@@ -104,7 +104,19 @@ func InfoOAuth(ctx *context.Context) {
 		Picture:  ctx.Doer.AvatarLink(ctx),
 	}
 
-	groups, err := oauth2_provider.GetOAuthGroupsForUser(ctx, ctx.Doer)
+	// groups, err := oauth2_provider.GetOAuthGroupsForUser(ctx, ctx.Doer)
+	var token string
+	if auHead := ctx.Req.Header.Get("Authorization"); auHead != "" {
```
Comment (lunny) [inline routers/web/auth/oauth2_provider.go]: This code looks unrelated? This is a web router, I don't think there is a header `Authorization` will be sent.

Comment (marcellmars) [reply routers/web/auth/oauth2_provider.go]: This is called from the API route `/login/oauth/userinfo` and it's the only endpoint other than the response coming embedded in the `id_token` if the scope/claim comes as `groups`.

This was my attempt to get the token so I can check if the `public-only` scope is present.

In my tests, the token is captured inside the header if one performs a request like this:

```
curl -H "Authorization: Bearer **token**" http://gitea.local/login/oauth/userinfo
```

(Given that membership in the public group is made visible)

If there's more idiomatic way to capture the token inside `/web/auth/*` I would be happy to change that into more idiomatic.

Comment (marcellmars) [reply routers/web/auth/oauth2_provider.go]: i found the better approach:

`onlyPublicGroups := ctx.IsSigned && (ctx.FormString("private") == "" || ctx.FormBool("private"))`

File: routers/web/auth/oauth2_provider.go [hunk 4]
```diff
@@ -104,7 +104,8 @@ func InfoOAuth(ctx *context.Context) {
 		Picture:  ctx.Doer.AvatarLink(ctx),
 	}
 
-	groups, err := oauth2_provider.GetOAuthGroupsForUser(ctx, ctx.Doer)
+	onlyPublicGroups := ctx.IsSigned && (ctx.FormString("private") == "" || ctx.FormBool("private"))
```
Comment (lunny) [inline routers/web/auth/oauth2_provider.go]: ctx.FormString("private") == "" || ctx.FormBool("private")
It seems it's not right here? Maybe `ctx.FormString("private") == "" || !ctx.FormBool("private")`.

Comment (marcellmars) [reply routers/web/auth/oauth2_provider.go]: that was verbatim line for the private i copied from [routers/web/repo/repo.go#L573](https://github.com/go-gitea/gitea/blob/32456b6f314f993efdc65fc90248b6fd1a8d55ef/routers/web/repo/repo.go#L573)

but meanwhile there were some changes with the new public only approach introduced recently.

after those changes `oauth2_provider.GetOAuthGroupsForUser` had the search into the database so the `private` can be done via an option without the need to traverse orgs as i originally did.

i changed my code into this:

```go
	form := web.GetForm(ctx).(*forms.AuthorizationForm)
	onlyPublicGroups, _ := oauth2_provider.GrantAdditionalScopes(form.Scope).PublicOnly()
	groups, err := oauth2_provider.GetOAuthGroupsForUser(ctx, ctx.Doer, onlyPublicGroups)
	if err != nil {
		ctx.ServerError("Oauth groups for user", err)
		return
	}
	response.Groups = groups
```

but dancing around too many repos i ended up in a major mess with my pull request so i think i will just do another pull request which would try to clean it up.

**Comment by marcellmars** [comment]
the new PR with cleaner commits would make sense.
